File: scripts/urlencode.js
#!/usr/bin/env node

const readline = require('readline');

function urlencode(s) {
    let encoded = '';
    for (let i = 0; i < s.length; i++) {
        let char = s.charAt(i);
        if (/^[a-zA-Z0-9.~_-]$/.test(char)) {
            encoded += char;
        } else {
            encoded += '%' + char.charCodeAt(0).toString(16).toUpperCase();
        }
    }
    return encoded;
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question("Enter the string you want to URL encode: ", function(inputString) {
    let encodedString = urlencode(inputString);
    console.log(encodedString);
    rl.close();
});

File: scripts/urlencode.py
#!/usr/bin/env python3

def urlencode(s):
    encoded = ''
    for char in s:
        if char.isalnum() or char in ['.', '~', '-', '_']:
            encoded += char
        else:
            encoded += '%' + char.encode('utf-8').hex().upper()
    return encoded

def main():
    input_string = input("Enter the string you want to URL encode: ")
    encoded_string = urlencode(input_string)
    print(encoded_string)

if __name__ == "__main__":
    main()
